xtc/src/ledger.rs

```diff
@@ -304,6 +316,227 @@ pub async fn transfer_from(from: Principal, to: Principal, amount: Nat) -> TxRec
     ))
 }
 
+pub type UsedBlocks = HashSet<BlockHeight>;
+pub type UsedMapBlocks = HashMap<BlockHeight, BlockHeight>;
+
+#[query(name = "getBlockUsed")]
+fn get_block_used() -> &'static HashSet<u64> {
+    let ic = get_context();
+    ic.get::<UsedBlocks>()
+}
+
+#[query(name = "isBlockUsed")]
+fn is_block_used(block_number: BlockHeight) -> bool {
+    let ic = get_context();
+    ic.get::<UsedBlocks>().contains(&block_number)
+}
+
+#[query]
+fn get_map_block_used(block_number: BlockHeight) -> Option<&'static BlockHeight> {
+    let ic = get_context();
+    ic.get::<UsedMapBlocks>().get(&block_number)
+}
+
+const ICPFEE: Tokens = Tokens::from_e8s(10000);
+const MEMO_TOP_UP_CANISTER: u64 = 1347768404_u64;
+const LEDGER_CANISTER_ID: CanisterId = CanisterId::from_u64(2);
+
+#[update]
+pub async fn mint_by_icp(sub_account: Option<Subaccount>, block_height: BlockHeight) -> TxReceipt {
+    IsShutDown::guard();
+
+    let ic = get_context();
+    let caller = ic.caller();
+
+    crate::progress().await;
+
+    let response: Result<BlockRes, (Option<i32>, String)> =
+        call_with_cleanup(LEDGER_CANISTER_ID, "block_pb", protobuf, block_height).await;
+    let encode_block = match response {
+        Ok(BlockRes(res)) => match res {
+            Some(result_encode_block) => match result_encode_block {
+                Ok(encode_block) => encode_block,
+                Err(e) => {
+                    let storage = match Principal::from_text(e.to_string()) {
+                        Ok(p) => p,
+                        Err(_) => return Err(TxError::Other),
+                    };
+                    let storage_canister = match CanisterId::new(PrincipalId::from(storage)) {
+                        Ok(c) => c,
+                        Err(_) => return Err(TxError::Other),
+                    };
+                    let response: Result<BlockRes, (Option<i32>, String)> =
+                        call_with_cleanup(storage_canister, "get_block_pb", protobuf, block_height)
+                            .await;
+                    match response {
+                        Ok(BlockRes(res)) => match res {
+                            Some(result_encode_block) => match result_encode_block {
+                                Ok(encode_block) => encode_block,
+                                Err(_) => return Err(TxError::Other),
+                            },
+                            None => return Err(TxError::Other),
+                        },
+                        Err(_) => return Err(TxError::Other),
+                    }
+                }
+            },
+            None => return Err(TxError::Other),
+        },
+        Err(_) => return Err(TxError::Other),
+    };
```

Comment:
To be honest I'm scared of this nested match expressions, let's see if we can improve it.

for example you can use `response.map_err(|_| TxError::Other)?` to remove the high level match expression, also on an Option<T> you can use `.ok_or(TxError::Other)?`